### PR TITLE
Fix handling of unicode in affix removal (${s#?} and ${s%?})

### PR DESCRIPTION
**Pull request by asokoloski**
Fixes: #173 

Hi @andychu, this one turned out to be a bit tricker than I thought.  I needed to change func_fnmatch in libc.c to set the locale to "" (making it refer to the environment variables) -- otherwise it assumed LANG=C, which meant that fnmatch did not properly handle multibyte characters.  I'm not sure if this was the best way to deal with this.

Also, it turns out that iterating backwards over a UTF-8 string is not substantially harder than going forwards, so I did that, rather than iterating forwards and building up a list of indexes as was suggested by your comment above DoUnarySuffixOp.

I hope this is alright.  I spent more time checking edge cases so as not to embarrass myself again :)

### CODE REVIEW

**Comment by andychu** [comment]
Cool!  I cloned this and ran the tests, and it worked.

I will look at the code in more detail later.  But does this make any assumptions about the locale environment variables?

My thinking is that OSH should only support utf-8, because it's a lot simpler to get all the edge cases right.  Unfortunately that is totally at odds with how libc string handling works.

I need to document that more.  I put some more notes in `doc/unicode.sh` just now.

I think we need some more testing with different env variable settings because there were some problems with the `wcswidth()` wrapper in a Docker image.  That function also sets the locale.

----

Long story short: since this works I will likely merge it.  But it's making me think about the general unicode strategy and #366 again.  Glob and fnmatch are tricky because it's something we don't want to reimplement ourselves.  So this is likely the right strategy for those operations.

**Comment by andychu** [comment]
OK thinking about it again, I think this is is the right strategy except that we should always change the locale to utf-8 for this operation.  That is, all our invocations of fnmatch() should be insensitive to the environment and respect utf-8.

This is how wcswidth() works right now.

I think I can rig up the spec framework to expose this.  Right now it only runs with LANG as utf-8 -- see `MakeTestEnv` in `sh_spec.py`.  Yeah let me play with it.

----

At least, this is the external behavior I think makes sense... I'm open to suggestions.

Basically Oil is following newer languages like Go, Rust, Julia, Swift, etc.  -- utf-8 is primary.   Rather than Java, JavaScript, and Python, which have "multibyte characters".

**Comment by asokoloski** [comment]
Yeah, I was not sure -- I think you're right, just copying what wcswidth does is better than what I did -- it doesn't make sense to take LANG into account when the code all expects utf-8 and won't work otherwise.  I can change it, unless you meant that you want to do it -- just let me know.

**Comment by asokoloski** [comment]
Oh, and about using utf-8 everywhere -- is there a good way to set this around where osh starts, rather than on every call to fnmatch and wcswidth?  We don't want to override someone's preferred date/time representation, etc, of course.  Would it be reasonable to look at the user-provided locale, and just replace the character encoding with utf-8, but keep the language the same?  I don't have enough experience with locales to know if I'm just spouting nonsense here.

**Comment by andychu** [comment]
Yeah that's a good question.  I'm not sure either :)  Unicode in libc is confusing.

Let me look at the code in more detail.  I think I will merge it without the change, and then use that to improve the test framework to handle different LANG.  And what we learn from that will probably inform the code style.

**Comment by asokoloski** [comment]
Ok, sounds good, thanks!

**Comment by andychu** [comment]
Also based on the list of operations I have here,

https://github.com/oilshell/oil/blob/master/doc/unicode.md

I would expect that some tests in

```
test/spec.sh var-op-patsub
test/spec.sh glob
test/spec.sh extglob-match
```

would be failing.  I don't remember if that's because they actually work or because we didn't test them :)  

For patsub, we translate a glob pattern to a regex, then use `regexec()` to get the position of the match.  Because `fnmatch()` doesn't give you that.  I learned that there is a weird mismatch in shell because historically ksh did not use libc APIs!  It used its own stuff before there was even POSIX I think.

File: osh/string_ops.py [hunk 1]
```diff
@@ -61,36 +61,64 @@ def _CheckContinuationByte(byte):
...
+    i += length
   except IndexError:
     raise util.InvalidUtf8(INCOMPLETE_CHAR)
 
   return i
 
 
+def _PreviousUtf8Char(s, i):
+  """
+  Given a string and a byte offset, returns the position of the
```
Comment (andychu) [inline osh/string_ops.py]: OK this looks like a short and useful piece of code.

1) Can you write unit tests for it in `osh/string_ops_test.py` ?  I don't think the error cases are covered by the spec tests.

2) Is there a URL to explain this algorithm?  I think I get it but it's not obvious upon reading.

Here's a link I found through Googling, but the explanation is unclear and some people are quibbling with it:

https://stackoverflow.com/questions/22257486/iterate-backwards-through-a-utf8-multibyte-string

Comment (asokoloski) [reply osh/string_ops.py]: 1. Sure, sorry, didn't even consciously notice the test module.  Still learning my way around the code base :)

2. I can try to find another one, but when I was working on it I only referred to that stackoverflow post and the description on wikipedia.  I'll try to come up with a clear explanation in the code.

Comment (asokoloski) [reply osh/string_ops.py]: 0b9445adea807545bb1bc8f49ea5424db78c0c26 -- added tests for this function, as well as _NextUtf8Char, which lacked them.
1e611ed4514fec8e0afd4fe8d480e91bbcea2063 -- added comment to explain algorithm in _PreviousUtf8Char

**Comment by andychu** [comment]
OK excellent, that makes sense now!

For most changes, I rely on spec tests more than unit tests, since external behavior is more important.  But this one has some corner cases that made a unit test worthwhile.

**Comment by asokoloski** [comment]
:D  no worries, I agree about the testing.
